### PR TITLE
Add support of multiple methods during scan #440

### PR DESCRIPTION
**Pull request by MD-Levitan**
Add support of multiple methods during scan #440

# Landing a Pull Request (PR)

Long form explanations of most of the items below can be found in the [CONTRIBUTING](https://github.com/epi052/feroxbuster/blob/master/CONTRIBUTING.md) guide.

## Branching checklist
- [x] There is an issue associated with your PR (bug, feature, etc.. if not, create one)
- [x] Your PR description references the associated issue (i.e. fixes #123456)
- [x] Code is in its own branch
- [x] Branch name is related to the PR contents
- [x] PR targets main

## Static analysis checks
- [x] All rust files are formatted using `cargo fmt`
- [x] All `clippy` checks pass when running `cargo clippy --all-targets --all-features -- -D warnings -A clippy::mutex-atomic`
- [x] All existing tests pass

## Documentation
- [ ] New code is documented using [doc comments](https://doc.rust-lang.org/stable/rust-by-example/meta/doc.html)
- [ ] Documentation about your PR is included in the README, as needed

## Additional Tests
- [ ] New code is unit tested
- [ ] New code is integration tested, as needed
- [ ] New tests pass

### CODE REVIEW

**Comment by epi052** [comment]
Howdy, and thanks for the PR! With the holidays, it'll take a little while longer for me to review, but wanted to let you know I saw and appreciate you taking the time to put it in!

**Comment by epi052** [comment]
Hey there! I haven't looked at the implementation yet, nor have I pulled the branch and tested anything, just did an initial pass of things I know need to be changed in order to add a new cli option (the checklist below).  Most are complete, though there are a few in the "Still needs" section. 

I'll be taking a look at the implementation soon, thanks again! 

---

To check/notes for me:
- can i use lowercase methods 
- do post's have a way of manipulating the body/data 

Still needs:
- ferox-config.toml.example
  - [ ] add new value
- tests/test_banner.rs
  - [ ] add new test so that the new `BannerEntry` is displayed and examined under test
- Cargo.toml
  - [ ] bump version - minor release +1 - patch release = 0
- [documentation](https://epi052.github.io/feroxbuster-docs/docs/)
  - [ ] update [example config file section](https://epi052.github.io/feroxbuster-docs/docs/configuration/ferox-config-toml/)
  - [ ] update [help output section](https://epi052.github.io/feroxbuster-docs/docs/configuration/command-line/)
  - [ ] add an [example](https://epi052.github.io/feroxbuster-docs/docs/examples/)
  - [ ] update [comparison table](https://epi052.github.io/feroxbuster-docs/docs/compare/)


Complete:
- config/container.rs
  - [x] add value to `Configuration` struct
  - [x] add value to `Configuration::default`
  - [x] update `Configuration` docstring
  - [x] add value to `Configuration::parse_cli_args`
  - [x] add value to `Configuration::merge_config`
- config/tests.rs
  - [x] update `Configuration::tests::setup_config_test` with new value
  - [x] update `Configuration::tests::default_configuration` with new value
  - [x] add test to check config parsing
- banner/container.rs
  - [x] add new `BannerEntry` to `Banner` struct
  - [x] update `Banner::new` with a `BannerEntry`
  - [x] update `print_to` with whether to print or not
- parser.rs
  - [x] add the option
- scan_manager/tests.rs
  - [x] add serialized version of new option to the feroxstates_feroxserialize_implementation test

**Comment by MD-Levitan** [comment]
Hi, according to your questions/notes, i want to add:
1. You can use lowercase, uppercase and almost everything. I used [from_bytes](https://docs.rs/reqwest/0.9.18/reqwest/struct.Method.html#method.from_bytes) method and it converts bytes to almost any method. But default methods are in upper case.
2. It's my fault, i forgot about data. So now it's just change HTTP method and send request with empty body. So I guess I can add `--data` param to  manipulate body.

Looking forward to your response soon.

**Comment by epi052** [comment]
> 2. It's my fault, i forgot about data. So now it's just change HTTP method and send request with empty body. So I guess I can add `--data` param to  manipulate body.

Not a fault at all! Tbh, I'm not even entirely sure I care about it really. 

My goal has always been content discovery, not fuzzing. The question becomes, does allowing a `data` param to control a  POST body further that goal? I'd love to hear your thoughts. Also, I may consult the hivemind and see what twitter has to say about it.

**Comment by epi052** [comment]
Looks like `--data` is something folks expect/want with POST requests. If you feel up to it, let's add it to this PR, if not, I can review yours and branch once it's in a good spot to add --data. Just lmk what you wanna do. I'm going to start reviewing the implementation this evening and expect to have it done tomorrow (as a heads up).  Thanks again!

**Comment by epi052** [comment]
# Notes (work in progress)

## wildcard for other methods 

may need to examine wildcard filtering and see if we can still filter known wildcards even when methods differ
```
200     POST       83l      333w     3526c https://bitdiscovery.com/api/experiments/configurations
WLD      GET       83l      333w     3526c Got 200 for https://bitdiscovery.com/api/a59255bc65b049fe8cb8405422f60a48 (url length: 32)
WLD         -         -         - Wildcard response is static; auto-filtering 3526 responses; toggle this behavior by using --dont-filter
WLD      GET       83l      333w     3526c Got 200 for https://bitdiscovery.com/api/a9c8fd7f4e4d45f1a3aaf4aa82b9e89b11762d96b0ca470aa2948b8f9dac40d2571b9b1b4ac1412da9b1bf36a666a346 (url length: 96)
200     HEAD        0l        0w        0c https://bitdiscovery.com/api/.hta
200     HEAD        0l        0w        0c https://bitdiscovery.com/api/.swf
200     HEAD        0l        0w        0c https://bitdiscovery.com/api/.perf
200     HEAD        0l        0w        0c https://bitdiscovery.com/api/.cvs
```

## methods not taken into account when determining scan length 

We know how many methods need to be scanned, so each additional one is a full iteration of the wordlist. With a wordlist of 4658 words, we should see an increase of 4658 for each additional method. We do the same thing for extensions, just need to add similar logic for methods.
```
wc -l /wordlists/seclists/Discovery/Web-Content/common.txt
4658 
```
```
[#####>--------------] - 17s     6648/23290   44s     found:20      errors:0      
[######>-------------] - 17s     1619/4658    94/s    https://bitdiscovery.com
[####>---------------] - 15s     1114/4658    73/s    https://bitdiscovery.com/api
[#>------------------] - 6s       300/4658    45/s    https://bitdiscovery.com/css
```

**Comment by MD-Levitan** [comment]
Thank you for review. I'll update my PR  - add `--data` and fix all  flaws.
But it will take a time for me to all these stuff. I hope that in next week I'll find time for it.

**Comment by MD-Levitan** [comment]
I have an update about wildcard. I rewrite this block with wildcard and now it trying to find wildcard for every HTTP method in config. But i guess it can be improved and create a HashMap with request and after compare the result for each method. What do you think about it?

Now it works in such way:

```bash
 🎯  Target Url            │ http://127.0.0.1:8000
 🚀  Threads               │ 50
 📖  Wordlist              │ /usr/share/seclists/Discovery/Web-Content/raft-medium-directories.txt
 👌  Status Codes          │ [200, 204, 301, 302, 307, 308, 401, 403, 405, 500]
 💥  Timeout (secs)        │ 7
 🦡  User-Agent            │ feroxbuster/2.4.2
 💉  Config File           │ /home/kusok/.config/feroxbuster/ferox-config.toml
 🏁  HTTP methods          │ [GET, PUT, POST]
 🔃  Recursion Depth       │ 4
 🎉  New Version Available │ https://github.com/epi052/feroxbuster/releases/latest
───────────────────────────┴──────────────────────
 🏁  Press [ENTER] to use the Scan Management Menu™
──────────────────────────────────────────────────
WLD      GET        1l        2w        9c Got 200 for http://127.0.0.1:8000/3450630f9c62434f961ba65a78bf058f (url length: 32)
WLD         -         -         - Wildcard response is static; auto-filtering 9 responses; toggle this behavior by using --dont-filter
WLD      GET        1l        2w        9c Got 200 for http://127.0.0.1:8000/288548d0f0814c0c9529c4e28b7e79797daf52028e364196a3542f5e600655885828d7d0b48c4d8ca8c1593549a29639 (url length: 96)
WLD      PUT        1l        3w       13c Got 200 for http://127.0.0.1:8000/28ddd1c81a444f6aa2ce2b6c2c0f4410 (url length: 32)
WLD         -         -         - Wildcard response is static; auto-filtering 13 responses; toggle this behavior by using --dont-filter
WLD      PUT        1l        3w       13c Got 200 for http://127.0.0.1:8000/fec8be6245db402ba3fb16c80222d76bcfa2a6a36d4645008b415f2104e3a10337024e13933c4f219dd73b1312ac8d0a (url length: 96)
WLD     POST        1l        2w        9c Got 200 for http://127.0.0.1:8000/39e02610f71c4fc4920118c26b92fe58 (url length: 32)
WLD         -         -         - Wildcard response is static; auto-filtering 9 responses; toggle this behavior by using --dont-filter
WLD     POST        1l        2w        9c Got 200 for http://127.0.0.1:8000/0381e086df48456c99eef0a368dd898fbda259d2e65e48d7b9f6df8ee21993bf6ac8e31147524492af6056da592e8fc3 (url length: 96)
200      GET        1l        6w       32c http://127.0.0.1:8000/check
🚨 Caught ctrl+c 🚨 saving scan state to ferox-http_127_0_0_1:8000-1640868007.state ...
[>-------------------] - 3s      3335/89997   1m      found:6       errors:0      
[#>------------------] - 3s      5140/89997   1376/s  http://127.0.0.1:8000
```
Code of test server to reproduce:
```rust
/// [dependencies]
/// rocket = "0.5.0-rc.1"


#[macro_use]
extern crate rocket;
use rocket::http::Method;
use rocket::http::Status;
use rocket::request::Request;

#[get("/check")]
fn hello() -> &'static str {
    "Check API"
}

#[catch(404)]
fn general_not_found(req: &Request) -> (Status, &'static str) {
    match req.method() {
        Method::Put => (Status::Ok, "not found put"),
        Method::Options => (Status::Ok, "not found options"),
        _ => (Status::Ok, "not found"),
    }
}

#[launch]
fn rocket() -> _ {
    rocket::build()
        .register("/", catchers![general_not_found])
        .mount("/", routes![hello])
}


```

**Comment by epi052** [comment]
> Thank you for review. I'll update my PR - add `--data` and fix all flaws. But it will take a time for me to all these stuff. I hope that in next week I'll find time for it.

no worries, I'm still reviewing, and there's absolutely no rush (as far as I'm concerned) for any of this. I'll update this thread with everything we need to wrangle before we merge. 

It may look like a lot of stuff still needs done, but you nailed the PR, it's very well done. Thanks again!

File: Cargo.toml
Comment (epi052) [inline Cargo.toml:3]: Since this is a new feature, let's bump it to 2.5.0

File: src/config/container.rs
Comment (epi052) [inline src/config/container.rs:176]: the default from `config::utils` should be used here

```#[serde(default = "methods")]```

**Review by epi052 — CHANGES_REQUESTED** [review]
This doesn't include the wildcard stuff, but I looked over everything else. It's just a few minor requests.

**Comment by epi052** [comment]
Ok, the review for everything except the new wildcard stuff is in, and I've updated [this comment](https://github.com/epi052/feroxbuster/pull/441#issuecomment-1001099764) to track things inside and outside the actual review. 

for the wildcard stuff, ill take a look at your branch and think on it a bit. Thanks again!

**Comment by MD-Levitan** [comment]
I have one question about `data`. Should be it's usual String? Because i think it would be cool if it's possible to pass data as binary string. I mean something like this: `\x10\x22\x33\...\x01`.
And now the main problem to do it, it's what `clap` doesn't support such input (at least i didn't fin it). So i see only one way it's to write parser for string.

**Comment by epi052** [comment]
> I have one question about `data`. Should be it's usual String? Because i think it would be cool if it's possible to pass data as binary string. I mean something like this: `\x10\x22\x33\...\x01`. And now the main problem to do it, it's what `clap` doesn't support such input (at least i didn't fin it). So i see only one way it's to write parser for string.

What about having `--data some-string` be `Vec<u8>` and `--data @filename` be how you pass 'actual' binary data. 

ferox would then look at first char and if it's an `@`, read the file contents, otherwise, parse the string into Vec<u8>. This would be similar behavior to curl.

```
man curl
       -d, --data <data>
               If you start the data with the letter @, the rest should be a file name to read the data from, or - if you want curl to read the data from stdin. Posting data from a file named
              'foobar' would thus be done with -d, --data @foobar. When -d, --data is told to read from a file like that, carriage returns and newlines will be stripped  out.  If  you  don't
              want the @ character to have a special interpretation use --data-raw instead.
```

File: src/config/container.rs
Comment (epi052) [reply src/config/container.rs]: fixed

File: src/event_handlers/outputs.rs
Comment (epi052) [reply src/event_handlers/outputs.rs]: :tada:

**Comment by epi052** [comment]
hey, started another review with the new changes. Should have it done tomorrow. 

Afterward I'll take a look at the wilcard stuff and see what's going on

**Comment by epi052** [comment]
oh, can you merge main as well when you get a moment? a PR was merged yesterday.

**Review by epi052 — CHANGES_REQUESTED** [review]
Ok, reviewed everything and made notes where appropriate.  Some things are inline comments in the review, and some others are in the Notes and/or checklist comment. 

Thanks for your hard work on this, it's looking very nice!

**Comment by MD-Levitan** [comment]
Hi, I've updated rep and added all your changes. So now it leaves to add `doc` and  `some methods not logged in progress bar`.
- `some methods not logged in progress bar` - I don't understand what you mean. Maybe your mean what we are filtering incorrect headers? If yes, than I guess it's correct way to save only corrrect methods for `reqwest`.
- `docs` - I really don't understand what i should do

**Comment by epi052** [comment]
> Hi, I've updated rep and added all your changes. So now it leaves to add `doc` and `some methods not logged in progress bar`.
> 
> * `some methods not logged in progress bar` - I don't understand what you mean. Maybe your mean what we are filtering incorrect headers? If yes, than I guess it's correct way to save only corrrect methods for `reqwest`.
i'll see if i can track down where it's happening. for each request we send to the target, we should also update the progress bar by one. That's not happening the way it should. In the output i put in the notes, it looks as though half of the requests aren't updating the progress bar. 

> * `docs` - I really don't understand what i should do
don't worry about it, i'll handle the docs

File: src/event_handlers/outputs.rs
Comment (epi052) [inline src/event_handlers/outputs.rs:217]:
```suggestion
                            resp.method().as_str(),
```

File: src/banner/container.rs
Comment (epi052) [inline src/banner/container.rs:323]:
```suggestion
        }).replace("\n", " ")
        .replace("\r", "");
```

File: src/config/container.rs
Comment (epi052) [inline src/config/container.rs:584]:
```suggestion
            if let Some(stripped) = arg.strip_prefix('@') {
```

File: src/config/container.rs
Comment (epi052) [inline src/config/container.rs:586]:
```suggestion
                    std::fs::read(stripped).unwrap_or_else(|e| report_and_exit(&e.to_string()));
```

File: src/banner/container.rs
Comment (epi052) [inline src/banner/container.rs:324]:
```suggestion
        })
        .replace("\n", " ")
        .replace("\r", "");
```

**Comment by epi052** [comment]
@MD-Levitan thanks for all your work on this, it's looking great! 

I'm going to merge it, but am finishing up a PR of my own, so am going to wait until that gets merged as well to push out 2.5.0. It shouldn't be too long though. Thanks again!

**Comment by MD-Levitan** [comment]
No big deal. Thank you for helping me

**Pull request merged**
